### PR TITLE
feat(events): add filter rules for prefixEqualsIgnoreCase, suffixEqualsIgnoreCase, wildcard, and anythingBut* matches

### LINKED ISSUE #28462: (aws-events): EventPattern support for wildcard filters
### Describe the feature

As this whats-new: [Amazon EventBridge announces support for wildcard filters in rules](https://aws.amazon.com/about-aws/whats-new/2023/10/amazon-eventbridge-wildcard-filters-rules/?nc1=h_ls) 

### Use Case

We can set event pattern in web console:
```
{
  "detail-type": ["Step Functions Execution Status Change"],
  "resources": [{
    "wildcard": "arn:aws:states:us-east-1:111122223333:execution:xxx:yyy*"
  }],
  "source": ["aws.states"]
}
```
But can not create this rule by CDK:
```
const ruleState = new Rule(this, 'ListenStateStatusChange', {
      description: 'Rule for listen SFN state machine status change',
      eventPattern: {
        source: ['aws.states'],
        detailType: ['Step Functions Execution Status Change'],
        resources: [`{ "wildcard": "arn:${Aws.PARTITION}:states:${Aws.REGION}:${Aws.ACCOUNT_ID}:execution:xxx:yyy*" }`],
      },
    });
```

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.81.0

### Environment details (OS name and version, etc.)

linux

### PR DESCRIPTION
**Pull request by jaw111**
### Issue # (if applicable)

Closes #28462.

### Reason for this change

Add support for:
* Prefix matching while ignoring case
* Suffix matching while ignoring case
* Wildcard matching
* Anything-but matching on suffixes
* Anything-but matching using wildcards
* Anything-but matching while ignoring case

Extend anything-but matching on prefixes to support a list of prefix values.

### Description of changes

Added functions on `Match` class:
* `prefixEqualsIgnoreCase()`
* `suffixEqualsIgnoreCase()`
* `wildcard()`
* `anythingButSuffix()`
* `anythingButWildcard()`
* `anythingButEqualsIgnoreCase()`

Modified `anythingButPrefix()` to support rest parameters.

### Description of how you validated changes

Added unit and integration tests for new and changed functions.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by jaw111** [comment]
The build fails, which seems to be due to failing tests for `@aws-cdk/aws-scheduler-targets-alpha:test`.

It's not clear from the contributing guidelines what/how to fix.

**Comment by badmintoncryer** [comment]
@jaw111 Thank you for your contribution!

It appears that aws-events integ test has failed, which may indicate that this modification introduces a breaking change.

```sh
@aws-cdk-testing/framework-integ: Failed: /codebuild/output/src3594926594/src/github.com/aws/aws-cdk/packages/@aws-cdk-testing/framework-integ/test/aws-events/test/integ.rule.js
@aws-cdk-testing/framework-integ: Error: Some tests failed!
```

Could you please clarify the reason for failure?

**Comment by badmintoncryer** [comment]
Oh, it seems the integration test you updated this time has failed. Could you please update the snapshot for `integ.rule.ts`?

**Comment by jaw111** [comment]
I think I am missing something, the contributing guide says to run integration tests on the `.js` file, whilst I modified the `.ts` file.

The changes I made to the `.ts` file are not reflected into the `.js` file, and the `.js` files appear to be ignored by Git.

How are those created/updated?

File: packages/aws-cdk-lib/aws-events/lib/event-pattern.ts
Comment (jaw111) [inline packages/aws-cdk-lib/aws-events/lib/event-pattern.ts:124]: This will always output an array also when only a single value is supplied.

To avoid changes to existing rules created with this pattern, I can add an exception to handle such cases.

**Comment by badmintoncryer** [comment]
It seems to be working well, but let me share the procedure just in case.

1. modify `integ.rule.ts`
2. execute `yarn tsc` at `aws-cdk/packages/@aws-cdk-testing/framework-integ`
3. execute `yarn integ path/to/integ.rule.js --update-on-failed`

File: packages/aws-cdk-lib/aws-events/lib/event-pattern.ts
Comment (badmintoncryer) [inline packages/aws-cdk-lib/aws-events/lib/event-pattern.ts:119]: Changing an existing argument from a single `string` to `string[]` is a breaking change.
If we implement this, we would need to use feature flags, which would increase the complexity of the PR.
Please either avoid breaking changes as much as possible or modify the PR to use feature flags.

Example PRs:
- https://github.com/aws/aws-cdk/pull/31639
- https://github.com/aws/aws-cdk/pull/31475

Comment (jaw111) [reply packages/aws-cdk-lib/aws-events/lib/event-pattern.ts:133]: Is it a breaking change? If the function is called with a single value like `anythingButPrefix('foo')` then it still works as before, so it _is_ backwards compatible.

Otherwise I can make it something like this:

```ts
public static anythingButPrefix(prefix: string, ...values: string[]): string[] {
  values.unshift(prefix);

  // rest of code
}
```

Comment (jaw111) [reply packages/aws-cdk-lib/aws-events/lib/event-pattern.ts:133]: Another option would be to use function overloading:

```ts
public static anythingButPrefix(prefix: string): string[];
public static anythingButPrefix(...values: string[]): string[] {
  return this.anythingButConjunction('prefix', values);
}
```

Comment (badmintoncryer) [reply packages/aws-cdk-lib/aws-events/lib/event-pattern.ts:133]: I'm really sorry for my misunderstandings.. This is not a breaking change and you don't need to update any more.

Comment (jaw111) [reply packages/aws-cdk-lib/aws-events/lib/event-pattern.ts:133]: No problem, thanks for taking the time to review.

What are the next step? Should I merge/rebase the latest changes from the `main` branch?

**Review by badmintoncryer — CHANGES_REQUESTED** [review]
In this revision, I believe not only the `wildcard` but also matches like `equals-ignore-case` have been newly added. Could you please make the following adjustments accordingly?

- Modify the PR title and description
- Add usage examples for all newly added public static methods in the README

**Review by badmintoncryer — APPROVED** [review]
Thanks!

**Comment by jaw111** [comment]
@badmintoncryer @mrgrain what are the next steps, is anything still needed from me?

**Comment by badmintoncryer** [comment]
@jaw111 Could you please merge the main branch? Since this PR has passed community review, the `needs-maintainer-review' label will be attached and the maintainer would review this.

**Comment by jaw111** [comment]
I merged the main branch, but no `needs-maintainer-review` label has been added...

File: packages/aws-cdk-lib/aws-events/README.md
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-events/README.md:148]:
```suggestion
The "anything but" matchers allow you to specify multiple arguments. For example:
```

File: packages/@aws-cdk-testing/framework-integ/test/aws-events/test/integ.rule.ts
Comment (gracelu0) [reply packages/@aws-cdk-testing/framework-integ/test/aws-events/test/integ.rule.ts:44]: why does updating the resource fail?

**Review by gracelu0 — DISMISSED** [review]
A few minor comments, otherwise LGTM! Once addressed I'll approve.

File: packages/@aws-cdk-testing/framework-integ/test/aws-events/test/integ.rule.ts
Comment (jaw111) [reply packages/@aws-cdk-testing/framework-integ/test/aws-events/test/integ.rule.ts:44]: See [wildcards and event pattern complexity](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-create-pattern-operators.html#eb-filtering-wildcard-matching-complexity).

> If a rule is too complex, EventBridge returns an `InvalidEventPatternException` when attempting to create the rule.

I'm not sure how the complexity of a rule is measured, nor where the limit is where a rule become "too complex", but adding the wildcard match to the existing rule was sufficient to push it over the limit causing the test to fail.

**Review by gracelu0 — APPROVED** [review]
Thank you for contributing!

**Comment by ScottRobinson03** [comment]
@gracelu0 Looks like the codecov check is broken here too, which is preventing this PR from merging (and because of the queue it's also preventing other PRs from merging).

**Pull request merged**
